include/exadg/operators/generalized_laplace_operator.h

```diff
@@ -0,0 +1,849 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2023 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_GENERALIZED_LAPLACE_OPERATOR_H_
+#define INCLUDE_EXADG_OPERATORS_GENERALIZED_LAPLACE_OPERATOR_H_
+
+#include <exadg/functions_and_boundary_conditions/evaluate_functions.h>
+#include <exadg/grid/grid_utilities.h>
+#include <exadg/operators/interior_penalty_parameter.h>
+#include <exadg/operators/operator_base.h>
+#include <exadg/operators/variable_coefficients.h>
+#include <exadg/poisson/user_interface/boundary_descriptor.h>
+#include <variant>
+
+namespace ExaDG
+{
+namespace GeneralizedLaplace
+{
+/**
+ * Free function for a generic expression of the multiplication with coefficients of different
+ * ranks.
+ */
+template<typename T, int coefficient_rank, int dim, typename Number>
+static inline DEAL_II_ALWAYS_INLINE //
+  T
+  coeff_mult(dealii::Tensor<coefficient_rank, dim, Number> const & coeff, T const & x)
+{
+  if constexpr(coefficient_rank == 4)
+    return dealii::double_contract<2, 0, 3, 1>(coeff, x);
+  else
+    return coeff * x;
+}
+
+enum class CoefficientModel
+{
+  ConstantFunction,
+  VariableFunction
+};
+
+/**
+ * Collection of data for the generalized Laplace kernel.
+ */
+template<int dim>
+struct KernelData
+{
+  //! Internal penalty factor for the calculation of `tau`
+  double IP_factor{1.0};
+
+  //!
+  bool calculate_and_store_cell_based_face_coefficients{false};
+
+  //!
+  CoefficientModel coefficient_model;
+
+  //!
+  std::shared_ptr<dealii::Function<dim>> coefficient_function;
+};
+
+/**
+ * Generalized Laplace kernel. This class is responsible for providing the core pieces of
+ * information required to evaluate the Laplace operator, i.e. the fluxes, numerical parameters
+ * (like the internal penalty parameter), and the coefficients.
```

Comment:
"interior penalty" should be explained for this generalized Laplace operator. Does the operator support both continuous and discontinuous Galerkin discretizations?

Reply:
I've done the implementation with only DG in my head. So I'm not sure if there are any intricacies apart from the interior penalty that I missed about continuous Galerkin. What do you think?